src/Storages/ObjectStorage/StorageObjectStorage.cpp

```diff
@@ -60,7 +74,23 @@ StorageObjectStorage::StorageObjectStorage(
     metadata.setConstraints(constraints_);
     metadata.setComment(comment);
 
-    setVirtuals(VirtualColumnUtils::getVirtualsForFileLikeStorage(metadata.getColumns()));
+    auto file_iterator = StorageObjectStorageSource::createFileIterator(
+        configuration,
+        object_storage,
+        distributed_processing_,
+        context,
+        {}, // predicate
+        metadata.getColumns().getAll(), // virtual_columns
+        nullptr, // read_keys
+        {} // file_progress_callback
+    );
+
+    Strings paths;
+
+    if (checkIfHiveSettingEnabled(context, configuration->getTypeName()))
+        if (auto file = file_iterator->next(0))
+            paths = {file->getPath()};
```

Comment:
It's better to avoid creating file iterator if setting is disabled. And let's create a separate method called `getPathSample` that will contain the logic of creating a file iterator and reading first path.
Also, I think we can improve this place, because in `resolveSchemaAndFormat` we can actually already iterate over files. We can add one more argument to `resolveSchemaAndFormat`: `String & sample_path` and fill it during schema inference with the first path we see. And here we will create file iterator only if `sample_path` is empty (that will mean that schema inference wasn't used).
So, ideally it will look something like this:
```c++
...
String path_sample;
resolveSchemaAndFormat(columns, configuration->format, object_storage, configuration, format_settings, context, path_sample);
...
if (path_sample.empty() && getQuerySettings().hive_partitioning)
    path_sample = getPathSample(...);
setVirtuals(VirtualColumnUtils::getVirtualsForFileLikeStorage(metadata.getColumns(), path_sample));
```